data/transactions/transaction_test.go

```diff
@@ -121,10 +121,111 @@ func TestGoNonparticipatingWellFormed(t *testing.T) {
 	require.Error(t, err)
 }
 
+func TestAppCallCreateWellFormed(t *testing.T) {
+	feeSink := basics.Address{0x7, 0xda, 0xcb, 0x4b, 0x6d, 0x9e, 0xd1, 0x41, 0xb1, 0x75, 0x76, 0xbd, 0x45, 0x9a, 0xe6, 0x42, 0x1d, 0x48, 0x6d, 0xa3, 0xd4, 0xef, 0x22, 0x47, 0xc4, 0x9, 0xa3, 0x96, 0xb8, 0x2e, 0xa2, 0x21}
+	specialAddr := SpecialAddresses{FeeSink: feeSink}
+	curProto := config.Consensus[protocol.ConsensusCurrentVersion]
+	futureProto := config.Consensus[protocol.ConsensusFuture]
+	addr1, err := basics.UnmarshalChecksumAddress("NDQCJNNY5WWWFLP4GFZ7MEF2QJSMZYK6OWIV2AQ7OMAVLEFCGGRHFPKJJA")
+	require.NoError(t, err)
+	usecases := []struct {
+		tx            Transaction
+		spec          SpecialAddresses
+		proto         config.ConsensusParams
+		expectedError error
+	}{
+		{
+			tx: Transaction{
+				Type: protocol.ApplicationCallTx,
+				Header: Header{
+					Sender:     addr1,
+					Fee:        basics.MicroAlgos{Raw: 1000},
+					LastValid:  105,
+					FirstValid: 100,
+				},
+				ApplicationCallTxnFields: ApplicationCallTxnFields{
+					ApplicationID: 0,
+					ApplicationArgs: [][]byte{
+						[]byte("write"),
+					},
+				},
+			},
+			spec:  specialAddr,
+			proto: curProto,
+		},
+		{
+			tx: Transaction{
+				Type: protocol.ApplicationCallTx,
+				Header: Header{
+					Sender:     addr1,
+					Fee:        basics.MicroAlgos{Raw: 1000},
+					LastValid:  105,
+					FirstValid: 100,
+				},
+				ApplicationCallTxnFields: ApplicationCallTxnFields{
+					ApplicationID: 0,
+					ApplicationArgs: [][]byte{
+						[]byte("write"),
+					},
+					ExtraProgramPages: 0,
+				},
+			},
+			spec:  specialAddr,
+			proto: curProto,
+		},
+		{
+			tx: Transaction{
+				Type: protocol.ApplicationCallTx,
+				Header: Header{
+					Sender:     addr1,
+					Fee:        basics.MicroAlgos{Raw: 1000},
+					LastValid:  105,
+					FirstValid: 100,
+				},
+				ApplicationCallTxnFields: ApplicationCallTxnFields{
+					ApplicationID: 0,
+					ApplicationArgs: [][]byte{
+						[]byte("write"),
+					},
+					ExtraProgramPages: 3,
```

Comment:
To test extra pages, don't you need to try inserting a big program?  I'm somewhat surprised this is considered well formed at all, since it's trying to create an app without any program bytes. But maybe the empty program is allowed?

Reply:
I believe this is done in userBalance_test.go

Reply:
That only confirms that we can _encode_ accounts with large apps in them.  So we know the max account size is still bounded by MaxEncodedAccountDataSize.  But this doesn't test that we actually _accept_ the txn. So the code where you compare the incoming program size to the value determined by extra pages is never exercised.

I'm not even sure if we've tested your allocbounds work, to ensure we can encode txns with very large programs in them.